### DISCUSSION TITLE
Windows Undercurl Support

### DISCUSSION THREAD

**Discussion by rawhat** [wezterm/wezterm#4989] [Q&A]
### What Operating System(s) are you running on?

Windows

### WezTerm version

20240206-225705-4f123a46

### Ask your question!

I didn't want to either pollute the existing issues, or create a new one without fully understanding the situation.

I previously asked on one of the issues (https://github.com/wez/wezterm/issues/415#issuecomment-910441992) about undercurl support.  This was identified as an unsupported feature in ConPTY.

From what I can tell, the linked issue in the `microsoft/terminal` repository is now resolved?  🎉

My question is... does this unlock that functionality for WezTerm?  If so, would there need to be changes made?  I did try installing the new preview version of the MS Terminal, and did not see any change in wezterm.  Which may be expected, just noting it.

Just curious about this, since it'd be nice to not have to SSH into WSL to get this functionality.  Not really a big deal either way.

Thanks!

**Comment by rawhat** [comment]
It'd probably help to have done any investigation before creating this... I see in the code that it prefers the vendored `conpty.dll` in the `assets/windows/conhost` directory.  So I guess my question is answered that "yes, this probably needs to be updated for this to work".  But also seems like a minimal change 😄

**Reply by akthe-at** [reply]
I, as fellow windows plebian, would love this.

**Reply by wez** [reply]
Yeah, it sounds like the conpty binaries bits included in this repo need to be updated. It's a few hours work to dust off a windows dev environment and figure out how to build just those bits.  It's time consuming throw-away work, and unfortunately it is not the kind of thing where I would feel comfortable taking the binaries built by someone else on their machine and just shipping them out to thousands of users; there's an element of trust missing.

The ideal situation is covered by https://github.com/wez/wezterm/issues/1927 which would allow us to ship (ideally signed!) binaries produced directly by MS.

@DHowett is there something you folks could potentially do to help streamline this in the interim?  eg: perhaps the microsoft/terminal GH CI could stash some artifacts in your builds and I could just download them and vendor them in?  You can disclaim all stability/warranty etc.

**Reply by wez** [reply]
@DHowett graciously uploaded a nuget package to use as a basis for this! I've pulled out the x64 executables from it and updated `main` in e7fe7c07da2da071d75c88ff3ae7a3c0e74d4f1f. That means that wezterm's next nightly build (typically takes 30-40 minutes for the windows build to update) should have this update.

I'm semi-recklessly trusting that those bits will work just fine without further packaging/installer attention because I'm not going to be firing up my local windows box today.  @rawhat, @akthe-at, please try the nightly build in about an hour from now and let me know if it starts at all(!) and whether it addresses the issue with undercurl.  Thanks!

**Reply by akthe-at** [reply]
> @DHowett graciously uploaded a nuget package to use as a basis for this! I've pulled out the x64 executables from it and updated `main` in e7fe7c07da2da071d75c88ff3ae7a3c0e74d4f1f. That means that wezterm's next nightly build (typically takes 30-40 minutes for the windows build to update) should have this update.
> 
> I'm semi-recklessly trusting that those bits will work just fine without further packaging/installer attention because I'm not going to be firing up my local windows box today.  @rawhat, @akthe-at, please try the nightly build in about an hour from now and let me know if it starts at all(!) and whether it addresses the issue with undercurl.  Thanks!
> 

Out walking the dog right now, can't wait to be a guinea pig when I get home, thanks @wez !

**Reply by rawhat** [reply]
Wow that was fast!  You're awesome 😃 

Looks great to me
![image](https://github.com/wez/wezterm/assets/2095509/31af4bc6-f603-4dde-9dbe-b9c9dffcaaf4)

Thanks so much for looking into this and addressing it so quickly.

**Reply by akthe-at** [reply]
@rawhat yours is showing up in Neovim??? What shell are you using??

**Reply by rawhat** [reply]
Yep!  I'm following this to get the terminfo https://wezfurlong.org/wezterm/faq.html?h=terminfo#how-do-i-enable-undercurl-curly-underlines

@wez has it all covered 😅

**Reply by akthe-at** [reply]
@rawhat Dang that doesn't seem to be working for me. Followed the neovim instructions and the 3 line script installed the wezterm terminfo file into my ~/.terminfo directory but no dice so far...

**Reply by rawhat** [reply]
Hmm, are you running it with `TERM=wezterm nvim` as well?  I think that's all I had to do 🤔

**Reply by akthe-at** [reply]
![image](https://github.com/wez/wezterm/assets/105152139/4c654543-e947-41be-913f-7c0cf7e96775)
![image](https://github.com/wez/wezterm/assets/105152139/c221d4e4-66c0-4ded-b14d-486554d3fba6)
The terminfo file path is C:/Users/ARK010/.terminfo/77/wezterm

**Reply by rawhat** [reply]
Oh I am also not sure exactly how to do this on Windows... sorry, misunderstood.  My above screenshot is in WSL, which does still benefit from these.

I'm definitely curious about getting it working outside of WSL too, though hmm

**Reply by akthe-at** [reply]
![image](https://github.com/wez/wezterm/assets/105152139/0fba0d53-3bf3-4e02-a139-570c29a5cd57)
It seems like it has to be possible? This is the stable wezterm from last week...You can see nothing happens with printf and undercurl. However, my first image has the nightly that @wez put up and it now works.

**Reply by akthe-at** [reply]
Using wezterm nightly, if i do `env TERM=tmux nvim` I get this as a result.
![image](https://github.com/wez/wezterm/assets/105152139/86bfcdab-2d3a-4092-9f2f-24b77d36480a)

If I use wezterm stable, and do the same, here is the result...

![image](https://github.com/wez/wezterm/assets/105152139/c6ee864a-afbd-4844-80a0-814fd55e45b7)

**Reply by Daydreamer-riri** [reply]
‌‌‌Hi @akthe-at, I'd like to ask how did you do it?

**Reply by akthe-at** [reply]
@Daydreamer-riri  If you don't do anything now, don't set the env var, neovim should now pick up on you using wezterm and show under curl properly. I stopped setting the env var to something else after they fixed that.

**Reply by Daydreamer-riri** [reply]
‌‌‌‌@akthe-at, I just used wezterm today, before that I was using Windows Terminal. The undercurl in Windows Terminal doesn't require any configuration.

I hope my wezterm also has undercurl, but the method in the [documentation](https://wezfurlong.org/wezterm/faq.html?h=under#how-do-i-enable-undercurl-curly-underlines) seems to be for Linux-like systems. I don't know what I should do.

**Reply by akthe-at** [reply]
> ‌‌‌‌@akthe-at, I just used wezterm today, before that I was using Windows Terminal. The undercurl in Windows Terminal doesn't require any configuration.
> 
> I hope my wezterm also has undercurl, but the method in the [documentation](https://wezfurlong.org/wezterm/faq.html?h=under#how-do-i-enable-undercurl-curly-underlines) seems to be for Linux-like systems. I don't know what I should do.

@Daydreamer-riri No configuration needed now!

**Reply by Daydreamer-riri** [reply]
![image](https://github.com/user-attachments/assets/4e38cdad-7db1-4660-9502-9f5b28c6d310)
‌‌‌‌‌

But I don't have undercurl now. 😭

**Reply by akthe-at** [reply]
@Daydreamer-riri what shell are you using? What wezterm version are you using? How are you testing to see if it works?

**Reply by Daydreamer-riri** [reply]
> @Daydreamer-riri what shell are you using? What wezterm version are you using? How are you testing to see if it works?

PowerShell. You can see the difference between the two.

**Reply by akthe-at** [reply]
@Daydreamer-riri and the wezterm version?

**Reply by Daydreamer-riri** [reply]
@akthe-at  It's  20240203-110809-5046fc22

Should I use the Nightly version?

**Reply by akthe-at** [reply]
@Daydreamer-riri I am using nightly but I thought that version should work? Regardless, if you copy this to the terminal outside of neovim does it work? 
```
printf "\x1b[58:2::255:0:0m\x1b[4:1msingle\x1b[4:2mdouble\x1b[4:3mcurly\x1b[4:4mdotted\x1b[4:5mdashed\x1b[0m\n"
```

**Reply by akthe-at** [reply]
,@Daydreamer-riri I don't remember if you need to modify that print test for powershell or not 
```
Write-Host "`e[58:2::255:0:0m`e[4:1msingle`e[4:2mdouble`e[4:3mcurly`e[4:4mdotted`e[4:5mdashed`e[0m"
```

**Reply by Daydreamer-riri** [reply]
‌‌I found that it works after I started using the Nightly version...

**Comment by akthe-at** [comment]
![image](https://github.com/wez/wezterm/assets/105152139/55dd38d4-ef17-4de8-a294-5c0b2a8bdd09)

Oh man oh man...we are cooking

**Reply by antoineco** [reply]
I just stumbled upon this comment, thanks for the heads up!

It even works in Neovim as of wezterm 20240513-190755-f1a08426:

![image](https://github.com/wez/wezterm/assets/3299086/e323e61e-5dc4-4e6f-873d-6daff23fe059)

**Reply by tienlonghungson** [reply]
> I just stumbled upon this comment, thanks for the heads up!
> 
> It even works in Neovim as of wezterm 20240513-190755-f1a08426:
> 
> ![image](https://private-user-images.githubusercontent.com/3299086/330799498-e323e61e-5dc4-4e6f-873d-6daff23fe059.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ2NzM5NzYsIm5iZiI6MTc0NDY3MzY3NiwicGF0aCI6Ii8zMjk5MDg2LzMzMDc5OTQ5OC1lMzIzZTYxZS01ZGM0LTRlNmYtODczZC02ZGFmZjIzZmUwNTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MTRUMjMzNDM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDM2NzA0ZjVhYmQwNmE2ZTA3MzQ2NWYxN2NiNWVlMDA5M2VmY2RiNzIyNGMwZGY2ODliMzc4MDUwZWRlY2FhYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.3M6-sDm6DWyHmSo7njCaDyss1YgfCcWaB0EV927K20U)

Hello, I am using Wezterm (installed through Winget). I open Nvim inside Wezterm, but undercurl does not work. How do you config it?

Running this "Write-Host "`e[58:2::255:0:0m`e[4:1msingle`e[4:2mdouble`e[4:3mcurly`e[4:4mdotted`e[4:5mdashed`e[0m"" in my Powershell returns "singledoublecurlydotteddashed"

**Reply by antoineco** [reply]
@tienlonghungson you need to install WezTerm _nightly_. The latest stable release is fairly old and doesn't contain this fix.
